test/langtools/jdk/javadoc/doclet/testSnippetTag/TestSnippetMarkup.java

```diff
@@ -0,0 +1,577 @@
+            String r = i.region.isBlank() ? "" : "region=" + i.region;
+            var methodDef = """
+
+                    /**
+                    {@snippet %s:
+                    %s}*/
+                    public void case%s() {}
+                    """.formatted(r, i.input, n);
+            methods.append(methodDef);
+        });
+        var classDef = """
+                public class A {
+                %s
+                }
+                """.formatted(methods.toString());
+        Path src = Files.createDirectories(base.resolve("src"));
+        tb.writeJavaFiles(src, classDef);
+        javadoc("-d", base.resolve("out").toString(),
+                "-sourcepath", src.toString(),
+                src.resolve("A.java").toString());
+        checkExit(Exit.OK);
+        checkNoCrashes();
+        forEachNumbered(testCases, (t, index) -> {
+            String html = """
+                        <span class="element-name">case%s</span>()</div>
+                        <div class="block">
+                        %s
+                        </div>""".formatted(index, getSnippetHtmlRepresentation("A.html", t.expectedOutput()));
+            checkOutput("A.html", true, html);
+        });
+    }
+
+    // FIXME: move error (i.e. negative) tests from TestSnippetTag to here
+
+    // @start region=<name> ... @end [region|region=<name>]
+    @Test
+    public void testStart(Path base) throws Exception {
+        var testCases = new ArrayList<TestCase>();
+        for (var variant : generateStartEndVariants()) {
+            var t = new TestCase(variant.region,
+                    """
+                            First line
+                              Second line // ###START
+                              Third line
+                              Fourth line // ###END
+                            Fifth line
+                            """.replaceFirst("###START", variant.start)
+                            .replaceFirst("###END", variant.end),
+                    """
+                            Second line
+                            Third line
+                            Fourth line""");
+            testCases.add(t);
+        }
+        testPositive(base, testCases);
+    }
+
+    private static String link(boolean linkPlain,
+                               String targetReference,
+                               String content)
+            throws UncheckedIOException {
+
+        // The HTML <a> tag generated from the @link snippet markup tag is the
+        // same as that of the {@link} Standard doclet tag. This is specified
+        // and can be used for comparison and testing.
+
+        // generate documentation for {@link} to grab its HTML <a> tag;
+        // generate documentation at low cost and do not interfere with the
+        // calling test state; for that, do not create file trees, do not write
+        // to std out/err, and generally try to keep everything in memory
```

Comment:
I thought about it and even half-implemented it before creating this PR. But a test purist in me demanded to have a clear way to get the link without interfering with the test logic. If it is not a showstopper, I would really like to integrate this PR to unlock more urgent work on the snippets feature before RDP 1, which is on 2021/12/09 as per the Schedule at https://openjdk.java.net/projects/jdk/18/. Sometime after this PR is integrated, we can revisit this.

Reply:
What you suggested is how I had that initially. But a test purist in me demanded a clear way of getting the link without interfering with the test logic.

As for text blocks: to me, indenting a text block is a complex, subjective and unsettled field. I would postpone re-indenting them for now, primarily because of the churn it would otherwise introduce to the dependent PRs. 

If neither of these issues is a showstopper, I would really like to integrate this PR to unlock more urgent work on the snippets feature before RDP 1, which is on 2021/12/09 as per the Schedule at https://openjdk.java.net/projects/jdk/18/. Sometime after this PR is integrated, we can revisit this.

Reply:
OK, but the test purist in me dislikes the significant amount of inline overhead to address this concern.

I'll pass on this for now, because I agree with unblocking the dependent PRs, but it seems like that is a negative side-effect of dependent PRs in general.  Once we get past this, at a minimum, I would suggest moving the in-memory file manager out of this class, and then potentially reconsider the decision that it is good practice to do the separate run.

Reply:
> As for text blocks: to me, indenting a text block is a complex, subjective and unsettled issue. I would postpone re-indenting them for now, primarily because of the churn it would otherwise introduce to the dependent PRs.

_complex, subjective and unsettled_?  Really? JEP 378 would indicate otherwise!   So, I disagree with the lack of indent, but (reluctantly) agree on the churn causing issues for dependent t PRs.

Reply:
(Cleanup suggestion, for later?)  maybe embed the `{@link}` in a separate comment, maybe even in a separate class or file, but still processed in the same run of `javadoc`.